src/client/mod.rs

```diff
@@ -571,199 +788,243 @@ impl ClientState {
-/// [`ClientExt`](./ext/trait.ClientExt.html) traits that provide methods of communicating with
-/// the server after connection. Cloning an `IrcClient` is relatively cheap, as it's equivalent to
-/// cloning a single `Arc`. This may be useful for setting up multiple threads with access to one
-/// connection.
-///
-/// For a full example usage, see [`irc::client`](./index.html).
-#[derive(Clone, Debug)]
-pub struct IrcClient {
-    /// The internal, thread-safe server state.
-    state: Arc<ClientState>,
-    /// A view of the logs for a mock connection.
-    view: Option<LogView>,
+/// Thread-safe sender that can be used with the client.
+#[derive(Debug, Clone)]
+pub struct Sender {
+    tx_outgoing: UnboundedSender<Message>,
 }
 
-impl Client for IrcClient {
-    fn config(&self) -> &Config {
-        &self.state.config
+impl Sender {
+    /// Send a single message to the unbounded queue.
+    pub fn send<M: Into<Message>>(&self, msg: M) -> error::Result<()> {
+        Ok(self.tx_outgoing.unbounded_send(msg.into())?)
     }
 
-    fn send<M: Into<Message>>(&self, msg: M) -> error::Result<()>
-    where
-        Self: Sized,
-    {
-        self.state.send(msg)
-    }
+    pub_state_base!();
+    pub_sender_base!();
+}
 
-    fn stream(&self) -> ClientStream {
-        ClientStream {
-            state: Arc::clone(&self.state),
-            stream: self.state.incoming.lock().unwrap().take().expect(
-                "Stream was already obtained once, and cannot be reobtained."
-            ),
+/// Future to handle outgoing messages.
+///
+/// Note: this is essentially the same as a version of [SendAll](https://github.com/rust-lang-nursery/futures-rs/blob/master/futures-util/src/sink/send_all.rs) that owns it's sink and stream.
+#[derive(Debug)]
+pub struct Outgoing {
+    sink: SplitSink<Connection, Message>,
+    stream: UnboundedReceiver<Message>,
+    buffered: Option<Message>,
+}
+
+impl Outgoing {
+    fn try_start_send(
+        &mut self,
+        cx: &mut Context<'_>,
+        message: Message,
+    ) -> Poll<Result<(), error::Error>> {
+        debug_assert!(self.buffered.is_none());
+
+        match Pin::new(&mut self.sink).poll_ready(cx)? {
+            Poll::Ready(()) => Poll::Ready(Pin::new(&mut self.sink).start_send(message)),
+            Poll::Pending => {
+                self.buffered = Some(message);
+                Poll::Pending
+            }
         }
     }
+}
 
-    #[cfg(not(feature = "nochanlists"))]
-    fn list_channels(&self) -> Option<Vec<String>> {
-        Some(
-            self.state
-                .chanlists
-                .lock()
-                .unwrap()
-                .keys()
-                .map(|k| k.to_owned())
-                .collect(),
-        )
+impl FusedFuture for Outgoing {
+    fn is_terminated(&self) -> bool {
+        // NB: outgoing stream never terminates.
+        // TODO: should it terminate if rx_outgoing is terminated?
+        false
```

Comment:
My impression is that the answer to this should be yes. It’d be ideal if sending a QUIT message in IRC actually ended the program execution if the client wasn’t set up to restart in a loop or something of that sort.

Reply:
Hm. Ending the execution is simply a matter of dropping the client and stream. I don't think QUIT messages are anything but informative, and the real way to quit a server is to disconnect. But I could be wrong.

Reply:
This is the client though? :thinking: